File: src/frontend/src/lib/api/user-profile.ts
import {
  mapCreateMyUserProfileResponse,
  mapGetMyUserProfileResponse,
  mapGetUserProfilesByPrincipalsRequest,
  mapGetUserProfilesByPrincipalsResponse,
  mapListStaleUsersResponse,
  mapListUserProfilesResponse,
  mapUpdateMyUserProfileRequest,
  mapUpdateUserProfileRequest,
  mapUserStatsResponse,
  type CreateMyUserProfileResponse,
  type GetMyUserProfileResponse,
  type GetUserProfilesByPrincipalsRequest,
  type GetUserProfilesByPrincipalsResponse,
  type ListStaleUsersResponse,
  type ListUserProfilesResponse,
  type UpdateMyUserProfileRequest,
  type UpdateUserProfileRequest,
  type UserProfile,
  type GetUserStatsResponse,
  mapOkResponse,
} from '@/lib/api-models';
import { isNotNil } from '@/lib/nil';
import type { ActorSubclass } from '@icp-sdk/core/agent';
import type { _SERVICE } from '@ssn/backend-api';

export class UserProfileApi {
  constructor(private readonly actor: ActorSubclass<_SERVICE>) {}

  public async getMyUserProfile(): Promise<GetMyUserProfileResponse> {
    const res = await this.actor.get_my_user_profile();

    return mapGetMyUserProfileResponse(res);
  }

  public async createMyUserProfile(): Promise<CreateMyUserProfileResponse> {
    const res = await this.actor.create_my_user_profile();

    return mapCreateMyUserProfileResponse(res);
  }

  public async getOrCreateMyUserProfile(): Promise<UserProfile> {
    const userProfile = await this.getMyUserProfile();

    if (isNotNil(userProfile)) {
      return userProfile;
    }

    return this.createMyUserProfile();
  }

  public async listUserProfiles(): Promise<ListUserProfilesResponse> {
    const res = await this.actor.admin_list_user_profiles();

    return mapListUserProfilesResponse(res);
  }

  public async updateMyUserProfile(
    req: UpdateMyUserProfileRequest,
  ): Promise<void> {
    const res = await this.actor.update_my_user_profile(
      mapUpdateMyUserProfileRequest(req),
    );
    mapOkResponse(res);
  }

  public async updateUserProfile(req: UpdateUserProfileRequest): Promise<void> {
    const res = await this.actor.admin_update_user_profile(
      mapUpdateUserProfileRequest(req),
    );
    mapOkResponse(res);
  }

  public async getUserStats(): Promise<GetUserStatsResponse> {
    const res = await this.actor.admin_get_user_stats();
    return mapUserStatsResponse(res);
  }

  public async listStaleUsers(): Promise<ListStaleUsersResponse> {
    const res = await this.actor.admin_list_stale_users();
    return mapListStaleUsersResponse(res);
  }

  public async verifyMyEmail(token: string): Promise<void> {
    const res = await this.actor.verify_my_email({ token });
    mapOkResponse(res);
  }

  public async getUserProfilesByPrincipals(
    req: GetUserProfilesByPrincipalsRequest,
  ): Promise<GetUserProfilesByPrincipalsResponse> {
    const res = await this.actor.get_user_profiles_by_principals(
      mapGetUserProfilesByPrincipalsRequest(req),
    );
    return mapGetUserProfilesByPrincipalsResponse(res);
  }
}

File: src/backend/src/controller/user_profile.rs
use crate::{
    dto::{
        CreateMyUserProfileResponse, GetMyUserProfileResponse, GetUserProfilesByPrincipalsRequest,
        GetUserProfilesByPrincipalsResponse, GetUserStatsResponse, ListUserProfilesResponse,
        UpdateMyUserProfileRequest, UpdateUserProfileRequest, VerifyEmailRequest,
    },
    service::{access_control_service, user_profile_service},
};
use canister_utils::{assert_authenticated, assert_controller, ApiResultDto};
use ic_cdk::{api::msg_caller, *};

#[query]
fn list_user_profiles() -> ApiResultDto<ListUserProfilesResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_controller(&caller) {
        return ApiResultDto::Err(err);
    }

    ApiResultDto::Ok(user_profile_service::list_user_profiles())
}

#[query]
fn get_user_profiles_by_principals(
    request: GetUserProfilesByPrincipalsRequest,
) -> ApiResultDto<GetUserProfilesByPrincipalsResponse> {
    let caller = msg_caller();
    if let Err(err) = access_control_service::assert_has_platform_access(&caller) {
        return ApiResultDto::Err(err);
    }

    user_profile_service::get_user_profiles_by_principals(&caller, request).into()
}

#[update]
fn update_user_profile(req: UpdateUserProfileRequest) -> ApiResultDto {
    let caller = msg_caller();
    if let Err(err) = assert_controller(&caller) {
        return ApiResultDto::Err(err);
    }

    user_profile_service::update_user_profile(req).into()
}

#[query]
fn get_my_user_profile() -> ApiResultDto<GetMyUserProfileResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    ApiResultDto::Ok(user_profile_service::get_my_user_profile(caller))
}

#[update]
fn create_my_user_profile() -> ApiResultDto<CreateMyUserProfileResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    user_profile_service::create_my_user_profile(caller).into()
}

#[update]
fn update_my_user_profile(req: UpdateMyUserProfileRequest) -> ApiResultDto {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    user_profile_service::update_my_user_profile(caller, req).into()
}

#[query]
fn get_user_stats() -> ApiResultDto<GetUserStatsResponse> {
    let caller = msg_caller();
    if let Err(err) = assert_controller(&caller) {
        return ApiResultDto::Err(err);
    }

    ApiResultDto::Ok(user_profile_service::get_user_stats())
}

#[update]
fn verify_my_email(req: VerifyEmailRequest) -> ApiResultDto {
    let caller = msg_caller();
    if let Err(err) = assert_authenticated(&caller) {
        return ApiResultDto::Err(err);
    }

    user_profile_service::verify_email(caller, req).into()
}
